docs/hooks/new-language.rst

```diff
@@ -1,105 +1,120 @@
+   -  Passes the value of the ``data`` key of the received object to the executed hook function.
+   -  Allows the hook function to modify the data.
 
-   -  When function was executed
+-  When a hook function is done:
+
+   -  Takes the modified data and serializes it back to JSON with the same ``uuid`` as it has received
+   -  Sends the JSON back as a TCP message
+   -  Sends a newline character ``\n`` as TCP message delimiter
+
+
+.. _tcp-socket-message-format:
+
+TCP socket message format
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+-  transaction (object)
+
+   -  uuid: ``234567-asdfghjkl`` (string) - ID used for unique identification of the message on both server and client sides
+   -  event: ``event`` (enum) - :ref:`Hook type <types-of-hooks>`
+
+      -  beforeAll (string) - Signals the hooks handler to run the ``beforeAll`` hooks
+      -  beforeEach (string) - Signals the hooks handler to run the ``beforeEach`` and ``before`` hooks
+      -  beforeEachValidation (string) - Signals the hooks handler to run the ``beforeEachValidation`` and ``beforeValidation`` hooks
+      -  afterEach (string) - Signals the hooks handler to run the ``after`` and ``afterEach`` hooks
+      -  afterAll (string) - Signals the hooks handler to run the ``afterAll`` hooks
+
+   -  data (enum) - Data passed as an argument to the hook function
+
+      -  (object) - Single :ref:`transaction object <transaction>`
+      -  (array) - An array of :ref:`transaction objects <transaction>`, containing all transactions Dredd currently works with; sent for ``beforeAll`` and ``afterAll`` events
 
-      -  It should serialize message to JSON
-      -  Send the serialized message back to the socket with same ``uuid`` as received
-      -  Send a newline character as message delimiter
 
 Termination
------------
+^^^^^^^^^^^
 
-When the testing is done, Dredd signals the hook handler process to terminate. This is done repeatedly with delays. When termination timeout is over, Dredd loses its patience and kills the process forcefully.
+When the testing is done, Dredd signals the hooks handler process to terminate. This is done repeatedly with delays. When termination timeout is over, Dredd loses its patience and kills the process forcefully.
 
 -  **retry delays** can be configured by :option:`--hooks-worker-term-retry`
 -  **timeout** can be configured by :option:`--hooks-worker-term-timeout`
 
-On Linux or macOS, Dredd uses the ``SIGTERM`` signal to tell the hook handler process it should terminate. On Windows, where signals do not exist, Dredd sends the ``END OF TEXT`` character (``\u0003``, which is ASCII representation of Ctrl+C) to standard input of the process.
+On Linux or macOS, Dredd uses the ``SIGTERM`` signal to tell the hooks handler process it should terminate. On Windows, where signals do not exist, Dredd sends the ``END OF TEXT`` character (``\u0003``, which is ASCII representation of Ctrl+C) to standard input of the process.
```

Comment:
This is hex `03`, I would prefer that or `\x03` instead of the unicode representation (it's simpler to understand).

Reply:
Again, this is one of those comments where I am not reviewing the changeset but how it already works (so this is to discuss/ -> new issue).

Could it be possible to close the TCP server socket or stdout stream instead on Windows? I.e, (`man 2 close`), (from my VERY VERY distant memory, I think this exists on Windows). Perhaps it would be simpler than the end of text, as calling `read` on a closed file descriptior would result in end-of-file.

Reply:
I'd appreciate if you could create a separate issue from the following:

> Could it be possible to close the TCP server socket or stdout stream instead on Windows? I.e, (man 2 close), (from my VERY VERY distant memory, I think this exists on Windows). Perhaps it would be simpler than the end of text, as calling read on a closed file descriptior would result in end-of-file.

TBH I thought about removing the Ctrl+C attempts altogether from Dredd. It is trying to do something which is not natural to the Windows platform and I think generally the ecosystem perhaps just expects processes to be killed without warning. Dredd is trying to be hyper-correct with sending `03` and I don't think any program will ever really understand that. I like your suggestion as an alternative, but currently process handling and the TCP socket are completely decoupled from each other in Dredd.